src/groupeddatatable/grouping.jl

```diff
@@ -232,21 +232,31 @@ dt = DataTable(a = repeat([1, 2, 3, 4], outer=[2]),
                b = repeat([2, 1], outer=[4]),
                c = randn(8))
 colwise(sum, dt)
+colwise([sum, lenth], dt)
+colwise((minimum, maximum), dt)
 colwise(sum, groupby(dt, :a))
 ```
 
 """
-colwise(f::Function, d::AbstractDataTable) = Any[vcat(f(d[idx])) for idx in 1:size(d, 2)]
-colwise(f::Function, gd::GroupedDataTable) = map(colwise(f), gd)
-colwise(f::Function) = x -> colwise(f, x)
-colwise(f) = x -> colwise(f, x)
+function colwise(f, d::AbstractDataTable)
+    x = [f(d[i]) for i in 1:ncol(d)]
+    if eltype(x) <: Nullable
+        return NullableArray(x)
+    else
+        return x
+    end
+end
 # apply several functions to each column in a DataTable
-colwise{T<:Function}(fns::Vector{T}, d::AbstractDataTable) =
-    reshape(Any[vcat(f(d[idx])) for f in fns, idx in 1:size(d, 2)],
-            length(fns)*size(d, 2))
-colwise{T<:Function}(fns::Vector{T}, gd::GroupedDataTable) = map(colwise(fns), gd)
-colwise{T<:Function}(fns::Vector{T}) = x -> colwise(fns, x)
-
+function colwise(fns::Union{AbstractVector, Tuple}, d::AbstractDataTable)
+    x = [f(d[i]) for f in fns, i in 1:ncol(d)]
+    if eltype(x) <: Nullable
+        return NullableArray(x)
+    else
+        return x
+    end
+end
+colwise(f, gd::GroupedDataTable) = [colwise(f, g) for g in gd]
```

Comment:
I'm afraid we broke something here. The previous definition (above) used to return a `GroupApplied` object, now we return an array of arrays, losing all information about groups. See the example given in the docstring above: `colwise(sum, groupby(dt, :a))`.

Is there any reason not to revert to this?
```julia
colwise{T<:Function}(fns::Vector{T}, gd::GroupedDataTable) = map(colwise(fns), gd)
```

Reply:
If we broke something then we should definitely revert the behavior. Did you catch this in your own work? If we revert it, let's make sure to provide a good test case too

Reply:
No, I just saw this when reading this old PR again in the context of `aggregate`... The docstring contains a testcase, unfortunately it's not a doctest.

Reply:
adding this
```
colwise{T<:Function}(fns::Vector{T}, gd::GroupedDataFrame) = map(colwise(fns), gd)
```

back still doesn't get us back to a `GroupApplied` object. We could implement the pre-this-PR colwise just for the `colwise{T<:Function}(fns::Vector{T}, gd::GroupedDataFrame)` case and leave the current behavior in place for other `AbstractDataFrames`, but with the current behavior we don't actually lose all information about the groups. Yes, we don't return a `GroupApplied` object, but the `GroupedDataFrame` has `n` groups and the returned vector of vectors will have `n` vectors that map 1:1 onto the groups, so all of that information is still available (or able to be reconstructed) if the user wants it. I'm not really sure I have a good intuition about which behavior is more useful until I see some real life example use cases

Reply:
> We really need to design a consistent plan for all these grouping functions

+💯

Could you open a new issue in DataFrames and map out a framework for what your desired inputs and outputs will be? I think I'm unclear on what the goals are for what DataFrames will support and what should be left to Query, DataFramesMeta, and [SAC](https://github.com/andyferris/SAC.jl)

Reply:
I wish I knew... I'm working on it, but that's not an easy question.